Add documentation for VMware vSphere (#11)
appuio/openshift4-docs: docs/modules/ROOT/pages/how-tos/vsphere/pre-install-checklist.adoc

====

With the defined variables, the following DNS records are required:

|===
|Record |Type |Target 

|`api.ocp4-vs.appuio.ch` |A/CNAME |Console Load Balancer

|`api-int.ocp4-vs.appuio.ch` |A/CNAME |Console Load Balancer

|`*.apps.ocp4-vs.appuio.ch` |A/CNAME |Apps Load Balancer

|`console.ocp4-vs.appuio.ch` |CNAME |Apps Load Balancer

|`etcd-0.ocp4-vs.appuio.ch` |A |Master 1

|`etcd-1.ocp4-vs.appuio.ch` |A |Master 2

|`etcd-2.ocp4-vs.appuio.ch` |A |Master 3

|`_etcd-server-ssl._tcp.ocp4-vs.appuio.ch` |SRV |See https://docs.openshift.com/container-platform/4.4/installing/installing_vsphere/installing-vsphere-network-customizations.html#installation-dns-user-infra_installing-vsphere-network-customizations[documentation] for details.
|===

[WARNING]
====
By default, Windows DNS servers create a SRV record for every A record.
For the A records of `etcd`, make sure that the SRV records are not created (or delete them, if they were), so the hostnames can be enumerated correctly.
====

=== Certificates

Certificates for the following hostnames need to be created:

* `api.ocp4-vs.appuio.ch`
* `console.ocp4-vs.appuio.ch`
* `*.apps.ocp4-vs.appuio.ch` (wildcard certificate)

[WARNING]
====

> For the certificates to be valid in modern web browser, make sure the _Subject Alternative Name (SAN)_ contains the hostname of the certificate.

Author: I agree, but prefer OpenSSL :smile: 
```suggestion
For the certificates to be valid in modern web browser, make sure the _Subject Alternative Name (SAN)_ contains the hostname of the certificate.
See https://www.switch.ch/pki/manage/request/csr-openssl/[How to create a CSR with OpenSSL] for an example on how to do so.
```

> [[additional-ca]]

Reviewer: Might be worth to place the part about CA in its own admonition.

Author: I had it that way before, looked kinda odd.

Reviewer: How so?

Author: I did not use the title before. Added title, looks good now!

If the certificates are issued from your own private PKI, the RootCA certificate of your PKI needs to be added to the `install-config.yaml` as described in the https://docs.openshift.com/container-platform/4.4/networking/configuring-a-custom-pki.html#installation-configure-proxy_configuring-a-custom-pki[documentation]:
[source,bash]
----

> ...

Reviewer: ```suggestion
…
```

additionalTrustBundle: | 
    -----BEGIN CERTIFICATE-----
    <MY_TRUSTED_CA_CERT_1>
    -----END CERTIFICATE-----
    -----BEGIN CERTIFICATE-----
    <MY_TRUSTED_CA_CERT_2>
    -----END CERTIFICATE-----

> ...

Reviewer: ```suggestion
…
```

There is a character for this. Just placing three dots, typographically looks odd.

----
Multiple CA certificates can be added.
====

=== Load balancers

For accessing API and applications, the following load balancers are required:

> * Console load balancer

Reviewer 2: Maybe `API load balancer` would be a more accurate name?
If I understand correctly the console is served via the apps load balancer in OpenShift 4?

Author: You are absolutely right. This is a leftover from 3.11 doc.

  ** Targets: all 3 masters as well as the bootstrap node at the beginning of the installation
  ** Ports: 6443/tcp, 22623/tcp
  ** TLS passthrough (no TLS termination on the load balancer
  ** WebSockets allowed
* Apps load balancer
  ** Targets: all 3 Infra Nodes
  ** Ports: 80/tcp, 443/tcp
  ** TLS passthrough (optimally no TLS termination on the load balancer)

=== Communication

The following communication has to be guaranteed:

|===
|Source |Destination |Ports |Comment

|OpenShift |Console load balancer |6443/tcp, 22623/tcp |All nodes

|OpenShift |App load balancer |80/tcp, 443/tcp |All nodes

|OpenShift |Outgoing proxy |depending on proxy |All nodes

|OpenShift masters |Configured IdP |depending on IdP |Authentication is handled by components running on masters
|===

Possible further surrounding systems:

* Access to Git repositories from all worker nodes
* Access to databases and other applications from all worker nodes

=== Outgoing proxy

> The following URLs need to be whitelisted on the outgoing proxy, as described in the https://docs.openshift.com/container-platform/4.4/installing/install_config/configuring-firewall.html[documentation].

Reviewer: ```suggestion
The following URLs need to be whitelisted on the outgoing proxy, as described at https://docs.openshift.com/container-platform/4.4/installing/install_config/configuring-firewall.html[Configuring your firewall].
```

If the connection to the proxy is done via HTTPS, the CA certificate, which was used to sign the certificate of the proxy server, needs to be added to the `install-config.yaml`, as described <<additional-ca,above>>.

> See https://docs.openshift.com/container-platform/4.4/installing/installing_vsphere/installing-vsphere.html#installation-configure-proxy_installing-vsphere[Configuring the cluster-wide proxy during installation] for details.

Reviewer: Placing the part about the CA in its own admonition, allows you to give it a title. You can then reference that title in the link text.

Reviewer: https://asciidoctor.org/docs/user-manual/ holds an example of what I mean. Search for _Feeding the Werewolves_.

[[fn2]]^2^: The bootstrap node is a temporary machine that can be deleted after the installation.

[[fn3]]^3^: Number and dimensioning of the worker nodes is dependent on your workload.

[[fn4]]^4^: RHCOS = Red Hat CoreOS

== Network

=== Address and DHCP

The OpenShift VMs should get their IP address via DHCP. Create DHCP reservations with the specified MAC address of the VM.

=== DNS records

Several DNS records are required for OpenShift to work correctly.

The https://docs.openshift.com/container-platform/4.4/installing/installing_vsphere/installing-vsphere-network-customizations.html#installation-dns-user-infra_installing-vsphere-network-customizations[OpenShift documentation] lists the required records in detail.

To illustrate, the following variables are set:

[horizontal]
`<base_domain>:`:: `appuio.ch`
`<cluster_name>:`:: `ocp4-vs`

[CAUTION]
====
`<cluster_name>` must match the folder name in vSphere, in which all OpenShift VMs reside, for vSphere storage integration to work correctly.
====

With the defined variables, the following DNS records are required:

|===
|Record |Type |Target 

> |`api.ocp4-vs.appuio.ch` |A/CNAME |Console Load Balancer

Reviewer 2: ```suggestion
|`api.ocp4-vs.appuio.ch` |A/CNAME |API Load Balancer
```

> |`api-int.ocp4-vs.appuio.ch` |A/CNAME |Console Load Balancer

Reviewer 2: ```suggestion
|`api-int.ocp4-vs.appuio.ch` |A/CNAME |API Load Balancer
```
